academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1559 @@
+            table_description=self.baseline_table_description,
+            ignore_unknown_values=True,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def create_snapshot(self, release: PubMedRelease, **kwargs):
+        """Create a snapshot of main table as a backup just in case something happens when applying the upserts and deletes."""
+
+        if not release.year_first_run:
+            snapshot_table_id = bq_sharded_table_id(
+                self.cloud_workspace.project_id, self.bq_dataset_id, f"{self.bq_table_id}_snapshot", release.start_date
+            )
+            expiry_date = pendulum.now().add(days=self.snapshot_expiry_days)
+            success = bq_snapshot(
+                src_table_id=self.main_table_id, dst_table_id=snapshot_table_id, expiry_date=expiry_date
+            )
+            logging.info(f"Created snapshot table: {snapshot_table_id}")
+            set_task_state(success, kwargs["ti"].task_id, release=release)
+        else:
+            logging.info("Not required to create a snapshot of the table for this run.")
+
+    def download_updatefiles(self, release: PubMedRelease, **kwargs):
+        """
+        Download the updatefiles from PubMed's FTP server for this release.
+
+        Unable to do this in parallel due to limitations of their FTP server.
+        """
+
+        updatefiles_datafiles = [datafile for datafile in release.datafile_list if not datafile.baseline]
+
+        success = download_datafiles(
+            datafile_list=updatefiles_datafiles,
+            ftp_server_url=self.ftp_server_url,
+            ftp_port=self.ftp_port,
+            reset_ftp_counter=self.reset_ftp_counter,
+            max_download_retry=self.max_download_retry,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def upload_downloaded_updatefiles(self, release: PubMedRelease, **kwargs):
+        """Upload downloaded updatefiles files to GCS."""
+
+        datafiles_to_upload = [
+            datafile.download_file_path for datafile in release.datafile_list if not datafile.baseline
+        ]
+
+        success = gcs_upload_files(
+            bucket_name=self.cloud_workspace.download_bucket,
+            file_paths=datafiles_to_upload,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def transform_updatefiles(self, release: PubMedRelease, **kwargs):
+        """
+        Transform the *.xml.gz files downloaded from PubMed's FTP server into usable json-like files for BigQuery import.
+
+        This is a multithreaded and pulls the PubmedArticle records from the downloaded XML files.
+        """
+
+        # Get list of datafiles for the baseline to trasform.
+        datafiles = [datafile for datafile in release.datafile_list if not datafile.baseline]
+
+        # Object to store all of the upserts and delete keys that are present in each file.
+        record_keys = {}
```

Comment:
Given the comments below, you would want to output a set of `PubmedUpdatefile` instances, then convert them to JSON to send them over XCom, then parse them again in the next task. It adds a bit of boilerplate but it is easier to work with (and for writing tests) and makes it easier to follow what is going on.